panels   = Channel.fromPath("${params.panel_dir}/*.bed")
samples  = Channel.fromPath("${params.cfDNA_dir}/*.vcf")

process score{
    tag { sample.baseName.replaceAll(/-called.vcf/,'') + " => " + panel.baseName.replaceAll(/_locations_hg19/,'') }
    echo true
    //memory '128 GB'
    executor 'local'
    //beforeScript "rm  -f ${params.home_dir}/detected_nofoci.tsv"
    afterScript "sort -k1,1 -n -r -o ${params.home_dir}/detected_nofoci.tsv ${params.home_dir}/foci_panel_noN.tsv"

    input:
    file panel from panels
    each sample from samples

    shell: 
    '''
    # Define some variables
    DIR="!{sample.getParent()}"
    PATIENT="!{sample.baseName.replaceAll(/-called.vcf/,'')}"
    PANEL="!{panel.baseName}"

    # Remove the chr prefix from chromosomes generated by freebayes
    !{params.home_dir}/../remove_chr.awk !{sample} > cfDNA.vcf

    # Remove common mutations from the sample (SKIP FOR NOW!)
    #bedtools subtract -header -a cfDNA.vcf -b !{params.common_variants} > all_somatic_cfDNA.vcf
    mv cfDNA.vcf all_somatic_cfDNA.vcf

    #Subset the panel
    head -n!{params.panel_size} !{params.panel_dir}/$PANEL.bed > panel.bed
    
    # Intersect focus somatic mutations and cfDNA panel variants
    bedtools intersect -a all_somatic_cfDNA.vcf -b panel.bed > somatic_cfDNA.vcf

    # Append output to the final results file
    echo -e `wc -l < somatic_cfDNA.vcf`'\t'`wc -l < cfDNA.vcf`'\t'!{sample}'\t'$PANEL >> !{params.home_dir}/foci_panel_noN.tsv
    '''
}
