panels   = Channel.fromPath("${params.panel_dir}/*.bed")
foci     = Channel.fromPath("${params.foci_dir}/P*.vcf").filter( ~/^.*P\d+-T.*/ ).map{file -> [((file =~ /(P\d+)/)[0][1]): file] }

process score{
    tag { focus[focus.keySet()[0]].baseName + " => " + panel.baseName.replaceAll(/_locations_hg19/,'') }
    echo true
    executor 'local'
    afterScript "sort -k5,5 -k2,2 -n -r ${params.home_dir}/somatic_cfDNA_panel.tsv > ${params.home_dir}/somatic_cfDNA_panel_sorted.tsv"

    input:
    file panel from panels
    each focus from foci

    shell: 
    '''
    # Define some variables
    DIR="!{focus[focus.keySet()[0]].getParent()}"
    PATIENT="!{focus.keySet()[0]}"
    FOCUS="!{focus[focus.keySet()[0]].baseName}"
    PANEL="!{panel.baseName}"

    # Remove germline mutations from focus
    bedtools subtract -header -a $DIR/$FOCUS.vcf -b $DIR/$PATIENT-N*.vcf > somatic.vcf
    
    # Remove the chr prefix from chromosomes generated by freebayes for cfDNA variants
    !{params.home_dir}/../remove_chr.awk !{params.cfDNA_dir}/$PATIENT-called.vcf > cfDNA.vcf

    #Subset the panel
    head -n!{params.panel_size} !{params.panel_dir}/$PANEL.bed | sort-bed - > panel.bed

    # Intersect focus somatic mutations and cfDNA variants
    bedtools intersect -a somatic.vcf -b cfDNA.vcf > somatic_cfDNA.vcf

    # Intersect cfDNA somatic mutations with the panel
    bedtools intersect -a somatic_cfDNA.vcf -b panel.bed > somatic_cfDNA_panel.vcf

    # Append output to the final results file
    echo -e `grep -v "#" somatic_cfDNA_panel.vcf | wc -l`'\t'`grep -v "#" somatic_cfDNA.vcf | wc -l`'\t'`grep -v "#" cfDNA.vcf | wc -l`'\t'`grep -v "#" somatic.vcf | wc -l`'\t'$FOCUS'\t'$PANEL >> !{params.home_dir}/somatic_cfDNA_panel.tsv
    '''
}
