foci     = Channel.fromPath("${params.foci_dir}/P*.vcf").filter( ~/^.*P\d+-T.*/ ).map{file -> [((file =~ /(P\d+)/)[0][1]): file] }

process score{
    tag { focus[focus.keySet()[0]].baseName }
    echo true
    executor 'local'
    afterScript "sort -k1,1 -n -r -o ${params.home_dir}/somatic_cfDNA.tsv ${params.home_dir}/somatic_cfDNA.tsv"

    input:
    each focus from foci

    shell: 
    '''
    # Define some variables
    DIR="!{focus[focus.keySet()[0]].getParent()}"
    PATIENT="!{focus.keySet()[0]}"
    FOCUS="!{focus[focus.keySet()[0]].baseName}"

    # Remove germline mutations from focus
    bedtools subtract -header -a $DIR/$FOCUS.vcf -b $DIR/$PATIENT-N*.vcf > somatic.vcf
    
    # Remove the chr prefix from chromosomes generated by freebayes for cfDNA variants
    !{params.home_dir}/../remove_chr.awk !{params.cfDNA_dir}/$PATIENT-called.vcf > cfDNA.vcf

    # Intersect focus somatic mutations and cfDNA variants
    bedtools intersect -a somatic.vcf -b cfDNA.vcf > somatic_cfDNA.vcf

    # Append output to the final results file
    echo -e `wc -l < somatic_cfDNA.vcf`'\t'`wc -l < somatic.vcf`'\t'$FOCUS >> !{params.home_dir}/somatic_cfDNA.tsv
    '''
}
